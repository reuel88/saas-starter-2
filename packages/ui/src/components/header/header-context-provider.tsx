"use client";

import { createContext, ReactNode, useContext } from "react";
import { User } from "../../types/user";

type HeaderContextType = {
  user: User | null;
};

export const HeaderContext = createContext<HeaderContextType>({
  user: null,
});

export const useHeader = () => {
  return useContext(HeaderContext);
};

type HeaderProviderProps = {
  children: ReactNode;
  user: User | null;
};

export const HeaderContextProvider = ({
  children,
  user,
}: HeaderProviderProps) => {
  return (
    <HeaderContext.Provider value={{ user }}>{children}</HeaderContext.Provider>
  );
};
