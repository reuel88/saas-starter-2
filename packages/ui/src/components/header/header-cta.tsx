"use client";

import { Button, buttonVariants } from "../button";
import { useHeader } from "./header-context-provider";
import { Popover, PopoverContent, PopoverTrigger } from "../ui/popover";
import { UserAvatar } from "../user-avatar";
import { LayoutDashboard, LogOut } from "lucide-react";

export const HeaderCta = () => {
  const { user } = useHeader();

  const haveFullName = user?.firstName && user?.lastName;
  const fullName = haveFullName
    ? `${user.firstName} ${user.lastName}`
    : user?.name || "";

  return (
    <>
      {!user && (
        <>
          <a
            href={`/sign-in`}
            className={buttonVariants({
              variant: "outline",
            })}
          >
            Sign in
          </a>
          <a href={`/sign-up`} className={buttonVariants()}>
            Get Started
          </a>
        </>
      )}

      {user && (
        <>
          <Popover>
            <PopoverTrigger className="rounded-full border">
              <UserAvatar
                firstName={user?.firstName}
                lastName={user?.lastName}
                name={user?.name}
                imageSrc={user?.avatar?.imageSrc}
              />
            </PopoverTrigger>
            <PopoverContent
              className="mt-2 space-y-2 divide-y p-2"
              side="bottom"
              align="end"
            >
              <div className="px-4">
                <p className="text-sm">{fullName}</p>
                <p className="text-sm text-muted-foreground">{user.email}</p>
              </div>

              <div className="pt-2">
                <a
                  href="/dashboard"
                  className={buttonVariants({
                    variant: "ghost",
                    className: "flex w-full items-center justify-between",
                  })}
                >
                  Dashboard <LayoutDashboard className="h-5 w-5" />
                </a>
              </div>

              <div className="pt-2">
                <Button
                  variant="ghost"
                  className="flex w-full items-center justify-between"
                >
                  Log out <LogOut className="h-5 w-5" />
                </Button>
              </div>
            </PopoverContent>
          </Popover>
        </>
      )}
    </>
  );
};
