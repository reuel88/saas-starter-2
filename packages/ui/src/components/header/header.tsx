"use server";

import { cn } from "@repo/utilities";
import { FC } from "react";
import { CompanyLogo } from "../company-logo";
import { Container } from "../container";
import { ThemeToggle } from "../theme-toggle";
import { Wrapper } from "../wrapper";
import { HeaderContextProvider } from "./header-context-provider";
import { HeaderCta } from "./header-cta";
import { User } from "../../types/user";

export type HeaderProps = {
  className?: string;
  user: User | null;
};
export const Header: FC<HeaderProps> = ({ className, user }) => {
  return (
    <HeaderContextProvider user={user}>
      <Wrapper as="header" className={cn("border-b", className)}>
        <Container className="flex justify-between gap-x-4 py-4">
          <h2 id="header-heading" className="sr-only">
            Header
          </h2>

          {/* Brand Logo */}
          <div>
            <a href="/">
              <CompanyLogo />
            </a>
          </div>

          {/* Nav Bar */}
          <div className="hidden flex-1 items-center justify-center lg:flex">
            {/* <NavBar menu={menu} /> */}
          </div>

          <div className="flex gap-x-4">
            {/* Theme Switcher */}
            <ThemeToggle />

            {/* Mobile Menu */}
            <div className="lg:hidden">{/* <MobileNav menu={menu} /> */}</div>

            {/* CTA */}
            <div className="hidden gap-x-4 lg:flex">
              <HeaderCta />
            </div>
          </div>
        </Container>
      </Wrapper>
    </HeaderContextProvider>
  );
};
