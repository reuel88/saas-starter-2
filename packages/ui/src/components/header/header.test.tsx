import { render, screen } from "@testing-library/react";
import { Header } from "./header";

describe("Header", () => {
  it("should render without user", () => {
    // Arrange

    // Act
    const { baseElement } = render(<Header user={null} />);

    // Assert
    expect(screen.getByText("Sign in")).toBeInTheDocument();
    expect(screen.getByText("Get Started")).toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });

  it("should render with user", () => {
    // Arrange
    const user = {
      name: "John Doe",
      email: "john@doe.com",
    };

    // Act
    const { baseElement } = render(<Header user={user} />);

    // Assert
    expect(screen.queryByText("Sign in")).not.toBeInTheDocument();
    expect(screen.queryByText("Get Started")).not.toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });
});
