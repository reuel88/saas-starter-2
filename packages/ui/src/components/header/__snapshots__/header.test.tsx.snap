// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > should render with user 1`] = `
<body>
  <div>
    <header
      class="bg-background text-foreground border-b"
    >
      <div
        class="mx-auto max-w-7xl px-6 lg:px-8 flex justify-between gap-x-4 py-4"
      >
        <h2
          class="sr-only"
          id="header-heading"
        >
          Header
        </h2>
        <div>
          <a
            href="/"
          >
            <div
              class="inline-block"
            >
              <span
                class="sr-only"
              >
                Blaiz Money
              </span>
              <svg
                class="lucide lucide-flame h-9 w-9 text-hue-500"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8.5 14.5A2.5 2.5 0 0 0 11 12c0-1.38-.5-2-1-3-1.072-2.143-.224-4.054 2-6 .5 2.5 2 4.9 4 6.5 2 1.6 3 3.5 3 5.5a7 7 0 1 1-14 0c0-1.153.433-2.294 1-3a2.5 2.5 0 0 0 2.5 2.5z"
                />
              </svg>
            </div>
          </a>
        </div>
        <div
          class="hidden flex-1 items-center justify-center lg:flex"
        />
        <div
          class="flex gap-x-4"
        >
          <button
            aria-expanded="false"
            aria-haspopup="menu"
            class="text-primary inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 w-10"
            data-state="closed"
            id="radix-:r2:"
            type="button"
          >
            <span
              class="sr-only"
            >
              Toggle theme
            </span>
            <svg
              class="lucide lucide-sun h-[1.2rem] w-[1.2rem] rotate-0 scale-100 transition-all dark:-rotate-90 dark:scale-0"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="4"
              />
              <path
                d="M12 2v2"
              />
              <path
                d="M12 20v2"
              />
              <path
                d="m4.93 4.93 1.41 1.41"
              />
              <path
                d="m17.66 17.66 1.41 1.41"
              />
              <path
                d="M2 12h2"
              />
              <path
                d="M20 12h2"
              />
              <path
                d="m6.34 17.66-1.41 1.41"
              />
              <path
                d="m19.07 4.93-1.41 1.41"
              />
            </svg>
            <svg
              class="lucide lucide-moon absolute h-[1.2rem] w-[1.2rem] rotate-90 scale-0 transition-all dark:rotate-0 dark:scale-100"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 3a6 6 0 0 0 9 9 9 9 0 1 1-9-9Z"
              />
            </svg>
          </button>
          <div
            class="lg:hidden"
          />
          <div
            class="hidden gap-x-4 lg:flex"
          >
            <button
              aria-controls="radix-:r4:"
              aria-expanded="false"
              aria-haspopup="dialog"
              class="rounded-full border"
              data-state="closed"
              type="button"
            >
              <span
                class="relative flex shrink-0 overflow-hidden rounded-full h-10 w-10"
              >
                <span
                  class="flex h-full w-full items-center justify-center rounded-full bg-muted"
                >
                  J
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
    </header>
  </div>
</body>
`;

exports[`Header > should render without user 1`] = `
<body>
  <div>
    <header
      class="bg-background text-foreground border-b"
    >
      <div
        class="mx-auto max-w-7xl px-6 lg:px-8 flex justify-between gap-x-4 py-4"
      >
        <h2
          class="sr-only"
          id="header-heading"
        >
          Header
        </h2>
        <div>
          <a
            href="/"
          >
            <div
              class="inline-block"
            >
              <span
                class="sr-only"
              >
                Blaiz Money
              </span>
              <svg
                class="lucide lucide-flame h-9 w-9 text-hue-500"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8.5 14.5A2.5 2.5 0 0 0 11 12c0-1.38-.5-2-1-3-1.072-2.143-.224-4.054 2-6 .5 2.5 2 4.9 4 6.5 2 1.6 3 3.5 3 5.5a7 7 0 1 1-14 0c0-1.153.433-2.294 1-3a2.5 2.5 0 0 0 2.5 2.5z"
                />
              </svg>
            </div>
          </a>
        </div>
        <div
          class="hidden flex-1 items-center justify-center lg:flex"
        />
        <div
          class="flex gap-x-4"
        >
          <button
            aria-expanded="false"
            aria-haspopup="menu"
            class="text-primary inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 w-10"
            data-state="closed"
            id="radix-:r0:"
            type="button"
          >
            <span
              class="sr-only"
            >
              Toggle theme
            </span>
            <svg
              class="lucide lucide-sun h-[1.2rem] w-[1.2rem] rotate-0 scale-100 transition-all dark:-rotate-90 dark:scale-0"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="4"
              />
              <path
                d="M12 2v2"
              />
              <path
                d="M12 20v2"
              />
              <path
                d="m4.93 4.93 1.41 1.41"
              />
              <path
                d="m17.66 17.66 1.41 1.41"
              />
              <path
                d="M2 12h2"
              />
              <path
                d="M20 12h2"
              />
              <path
                d="m6.34 17.66-1.41 1.41"
              />
              <path
                d="m19.07 4.93-1.41 1.41"
              />
            </svg>
            <svg
              class="lucide lucide-moon absolute h-[1.2rem] w-[1.2rem] rotate-90 scale-0 transition-all dark:rotate-0 dark:scale-100"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 3a6 6 0 0 0 9 9 9 9 0 1 1-9-9Z"
              />
            </svg>
          </button>
          <div
            class="lg:hidden"
          />
          <div
            class="hidden gap-x-4 lg:flex"
          >
            <a
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
              href="/sign-in"
            >
              Sign in
            </a>
            <a
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
              href="/sign-up"
            >
              Get Started
            </a>
          </div>
        </div>
      </div>
    </header>
  </div>
</body>
`;
